httpx/_types.py

```diff
@@ -38,7 +39,14 @@
 ]
 
 HeaderTypes = Union[
-    "Headers", Dict[StrOrBytes, StrOrBytes], Sequence[Tuple[StrOrBytes, StrOrBytes]],
+    "Headers",
+    # NOTE: Mapping is invariant in key (https://github.com/python/mypy/issues/1114).
+    # So using `StrOrBytes` as the mapping key type would result in
+    # users come across https://github.com/python/mypy/issues/8477.
+    # The suggested solution is to use a generic key type (defined as a `TypeVar`),
+    # hence `AnyStr` here.
+    Mapping[AnyStr, StrOrBytes],
+    Sequence[Tuple[StrOrBytes, StrOrBytes]],
```

Comment:
Okay, so let's dig into something here...

Should we just use `AnyStr` everywhere here instead?

If we're using `Mapping[AnyStr, AnyStr]` I get that it implies that the type of the key must be the same throughout, and that the type of the value must be the same throughout. Does it also imply that the key and value types must also match?

In any case, why *not* adopt an invariant constraint on the types a user can properly use here. Either "all bytes" or "all str" is likely to result in tighter, better code.

Reply:
Honestly not sure about this - maybe it would be okay to enforce « either str or bytes, not a mix of both », and have users encode/decode explicitly. But at the same time it would require some runtime checks to enforce this behavior (not just at the type hints level), yet the current runtime behavior has been working well enough it seems.

In any case I’m tempted to push forward for this fix as currently we have a UX bug regardless of what we decide about this. Thoughts?

Reply:
> Does it also imply that the key and value types must also match?

Apparently not.

```python
import typing

AnyStrMapping = typing.Mapping[typing.AnyStr, typing.AnyStr]
MustMatchMapping = typing.Union[
    typing.Mapping[str, str], typing.Mapping[bytes, bytes],
]
m1: AnyStrMapping = {b"foo": "bar"}  # ok
m2: MustMatchMapping = {"foo": "bar"}  # ok
m3: MustMatchMapping = {b"foo": b"bar"}  # ok
m4: MustMatchMapping = {b"foo": "bar"}  # mypy error
```

I'm 👍  on merging this for the time being and discuss this further in another issue (FWIW I prefer the must-match version).

Reply:
So I'd be okay with this *if* it switches to `AnyStr` throughout those *four* cases, rather than just the one key there. And at that point I think we can drop the comment too.

Alternately #999, which is similar, but also introduces some tighter checking on similar signatures in `_utils.py`